cps/kni.c

```diff
@@ -381,72 +381,121 @@ attr_get(struct cps_config *cps_conf,
+			if (cps_conf->debug)
+				inet_print("src", addr);
+		} else if (likely(family == AF_INET6)) {
 			struct in6_addr *addr =
 				mnl_attr_get_payload(tb[RTA_SRC]);
 			if (cps_conf->debug)
 				inet6_print("src", addr);
 		} else {
-			struct in_addr *addr =
-				mnl_attr_get_payload(tb[RTA_SRC]);
-			if (cps_conf->debug)
-				inet_print("src", addr);
+			RTE_LOG(WARNING, GATEKEEPER,
+				"cps update: unknown address family %d at %s!\n",
+				family, __func__);
+			return;
 		}
 	}
 	if (tb[RTA_OIF]) {
+		update->oif_index = mnl_attr_get_u32(tb[RTA_OIF]);
+
 		if (cps_conf->debug)
 			RTE_LOG(INFO, GATEKEEPER, "cps update: oif=%u\n",
-				mnl_attr_get_u32(tb[RTA_OIF]));
+				update->oif_index);
 	}
 	if (tb[RTA_FLOW]) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"cps update: the rtnetlink command has information (RTA_FLOW) that we don't need or don't honor!\n");
+
 		if (cps_conf->debug)
 			RTE_LOG(INFO, GATEKEEPER, "cps update: flow=%u\n",
 				mnl_attr_get_u32(tb[RTA_FLOW]));
 	}
 	if (tb[RTA_PREFSRC]) {
-		if (family == AF_INET6) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"cps update: the rtnetlink command has information (RTA_PREFSRC) that we don't need or don't honor!\n");
+
+		if (family == AF_INET) {
+			struct in_addr *addr =
+				mnl_attr_get_payload(tb[RTA_PREFSRC]);
+			if (cps_conf->debug)
+				inet_print("prefsrc", addr);
+		} else if (likely(family == AF_INET6)) {
 			struct in6_addr *addr =
 				mnl_attr_get_payload(tb[RTA_PREFSRC]);
 			if (cps_conf->debug)
 				inet6_print("prefsrc", addr);
 		} else {
-			struct in_addr *addr =
-				mnl_attr_get_payload(tb[RTA_PREFSRC]);
-			if (cps_conf->debug)
-				inet_print("prefsrc", addr);
+			RTE_LOG(WARNING, GATEKEEPER,
+				"cps update: unknown address family %d at %s!\n",
+				family, __func__);
+			return;
 		}
 	}
 	if (tb[RTA_GATEWAY]) {
-		if (family == AF_INET6) {
+		if (family == AF_INET) {
+			struct in_addr *addr =
+				mnl_attr_get_payload(tb[RTA_GATEWAY]);
+			rte_memcpy(&update->gw.v4,
+				addr, sizeof(update->gw.v4));
+			if (cps_conf->debug)
+				inet_print("gw", addr);
+		} else if (likely(family == AF_INET6)) {
 			struct in6_addr *addr =
 				mnl_attr_get_payload(tb[RTA_GATEWAY]);
+			rte_memcpy(&update->gw.v6,
+				addr, sizeof(update->gw.v6));
 			if (cps_conf->debug)
 				inet6_print("gw", addr);
 		} else {
-			struct in_addr *addr =
-				mnl_attr_get_payload(tb[RTA_GATEWAY]);
-			if (cps_conf->debug)
-				inet_print("gw", addr);
+			RTE_LOG(WARNING, GATEKEEPER,
+				"cps update: unknown address family %d at %s!\n",
+				family, __func__);
+			return;
 		}
+
+		gw_present = true;
 	}
 	if (tb[RTA_PRIORITY]) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"cps update: the rtnetlink command has information (RTA_PRIORITY) that we don't need or don't honor!\n");
+
 		if (cps_conf->debug)
 			RTE_LOG(INFO, GATEKEEPER, "cps update: prio=%u\n",
 				mnl_attr_get_u32(tb[RTA_PRIORITY]));
 	}
+
+	update->valid = dst_present &&
+		(update->type != RTM_NEWROUTE || gw_present);
```

Comment:
I know that I suggested the expression above, but both, the previous code and my expression, consider that `update->type` is either `RTM_NEWROUTE`, or `RTM_DELROUTE`. The following expression makes the robust test we need.
```
update->valid = dst_present &&
	(update->type == RTM_DELROUTE ||
	 update->type == RTM_NEWROUTE && gw_present);
```